pkg/trace/api/otlp.go

```diff
@@ -203,70 +203,136 @@ func fastHeaderGet(h http.Header, canonicalKey string) string {
-				EndpointVersion: fmt.Sprintf("opentelemetry_%s_v1", protocol),
-			},
-			Stats: info.NewStats(),
-		}
-		tracesByID := make(map[uint64]pb.Trace)
-		for i := 0; i < rspans.ScopeSpans().Len(); i++ {
-			libspans := rspans.ScopeSpans().At(i)
-			lib := libspans.Scope()
-			for i := 0; i < libspans.Spans().Len(); i++ {
-				span := libspans.Spans().At(i)
-				traceID := traceIDToUint64(span.TraceID().Bytes())
-				if tracesByID[traceID] == nil {
-					tracesByID[traceID] = pb.Trace{}
+		o.ReceiveResourceSpans(rspans, header, protocol)
+	}
+}
+
+// OTLPIngestSummary returns a summary of the received resource spans.
+type OTLPIngestSummary struct {
+	// Hostname indicates the hostname of the passed resource spans.
+	Hostname string
+	// Tags returns a set of Datadog-specific tags which are relevant for identifying
+	// the source of the passed resource spans.
+	Tags []string
+}
+
+// ReceiveResourceSpans processes the given rspans and sends them to writer.
+func (o *OTLPReceiver) ReceiveResourceSpans(rspans ptrace.ResourceSpans, header http.Header, protocol string) OTLPIngestSummary {
+	// each rspans is coming from a different resource and should be considered
+	// a separate payload; typically there is only one item in this slice
+	attr := rspans.Resource().Attributes()
+	rattr := make(map[string]string, attr.Len())
+	attr.Range(func(k string, v pcommon.Value) bool {
+		rattr[k] = v.AsString()
+		return true
+	})
+	hostname, _ := attributes.HostnameFromAttributes(attr)
+	if hostname == "" {
+		hostname = rattr["_dd.hostname"]
+	}
+	env := rattr[string(semconv.AttributeDeploymentEnvironment)]
+	lang := rattr[string(semconv.AttributeTelemetrySDKLanguage)]
+	if lang == "" {
+		lang = fastHeaderGet(header, headerLang)
+	}
+	containerID := rattr[string(semconv.AttributeContainerID)]
+	if containerID == "" {
+		containerID = rattr[string(semconv.AttributeK8SPodUID)]
+	}
+	if containerID == "" {
+		containerID = fastHeaderGet(header, headerContainerID)
+	}
+	tagstats := &info.TagStats{
+		Tags: info.Tags{
+			Lang:            lang,
+			LangVersion:     fastHeaderGet(header, headerLangVersion),
+			Interpreter:     fastHeaderGet(header, headerLangInterpreter),
+			LangVendor:      fastHeaderGet(header, headerLangInterpreterVendor),
+			TracerVersion:   fmt.Sprintf("otlp-%s", rattr[string(semconv.AttributeTelemetrySDKVersion)]),
+			EndpointVersion: fmt.Sprintf("opentelemetry_%s_v1", protocol),
+		},
+		Stats: info.NewStats(),
+	}
+	tracesByID := make(map[uint64]pb.Trace)
+	for i := 0; i < rspans.ScopeSpans().Len(); i++ {
+		libspans := rspans.ScopeSpans().At(i)
+		lib := libspans.Scope()
+		for i := 0; i < libspans.Spans().Len(); i++ {
+			span := libspans.Spans().At(i)
+			traceID := traceIDToUint64(span.TraceID().Bytes())
+			if tracesByID[traceID] == nil {
+				tracesByID[traceID] = pb.Trace{}
+			}
+			ddspan := o.convertSpan(rattr, lib, span)
+			if hostname == "" {
```

Comment:
ditto here, we should not assume that because `hostname` is empty it is missing, on serverless environments we don't have a hostname

Reply:
Done.